nemo/collections/nlp/data/machine_translation/preproc_mt_data.py

```diff
@@ -250,107 +272,67 @@ def preprocess_parallel_dataset(
-                    decoder_tokenizer=decoder_tokenizer,
-                    num_files_in_tar=num_files_in_tar,
-                    tar_file_ptr=tar_file_ptr,
-                    tar_file_ctr=tar_file_ctr,
-                    global_batch_ctr=global_batch_ctr,
-                    pkl_file_prefix=pkl_file_prefix,
-                )
-                tar_file_ptr.close()
-                os.remove(tmp_f_src.name)
-                os.remove(tmp_f_tgt.name)
+                filenames = [src_fname, tgt_fname]
 
-                if num_files_in_tar != num_batches_per_tarfile:
-                    os.remove(
-                        os.path.join(
-                            out_dir, '%s-batches.tokens.%d.%d.tar' % (pkl_file_prefix, tokens_in_batch, tar_file_ctr)
-                        )
+                # get number of lines so that we can create a partition of the lines of the text file
+                num_src_lines, num_tgt_lines = Parallel(n_jobs=2)(
+                    delayed(MTDataPreproc._get_num_lines)(filename) for filename in filenames
+                )
+                logging.info(f'Found {num_src_lines} source lines and  {num_tgt_lines} target lines.')
+                assert num_src_lines == num_tgt_lines, 'Number of source lines should equal number of target lines.'
+
+                # create a partition of lines that we can parallelize over
+                lines_partition = MTDataPreproc._get_lines_partition(num_src_lines, lines_per_dataset_fragment)
+
+                # create tarfiles for each fragment in parallel
+                # total_batches = 0
+                total_batches_list = Parallel(n_jobs=n_jobs)(
+                    delayed(MTDataPreproc._process_fragment)(
+                        src_filename=src_fname,
+                        tgt_filename=tgt_fname,
+                        lines_indices=lines_indices,
+                        out_dir=out_dir,
+                        num_batches_per_tarfile=num_batches_per_tarfile,
+                        clean=clean,
+                        max_seq_length=max_seq_length,
+                        min_seq_length=min_seq_length,
+                        tokens_in_batch=tokens_in_batch,
+                        encoder_tokenizer_name=encoder_tokenizer_name,
+                        encoder_tokenizer_model=encoder_tokenizer_model,
+                        encoder_bpe_dropout=encoder_bpe_dropout,
+                        encoder_model_name=encoder_model_name,
+                        decoder_tokenizer_name=decoder_tokenizer_name,
+                        decoder_tokenizer_model=decoder_tokenizer_model,
+                        decoder_bpe_dropout=decoder_bpe_dropout,
+                        decoder_model_name=decoder_model_name,
+                        pkl_file_prefix=fragment_index,
                     )
-                    global_batch_ctr -= num_files_in_tar
-                    logging.info('Dropping %d batches because of overflow' % (num_files_in_tar))
```

Comment:
You need to move this dropping logic of batches into `_process_fragment()` otherwise tarfiles create from a particular fragment may be imbalanced.

Reply:
I drop batches here: https://github.com/NVIDIA/NeMo/blob/146952ca71c98f2274c49f2472a080f786157e45/nemo/collections/nlp/data/machine_translation/preproc_mt_data.py#L729

Reply:
Updated the batch dropping logic. Now we only discard at most num_batches_per_tarfile